packages/signing-utils/src/ssh-client.ts

```diff
@@ -0,0 +1,106 @@
+import type { ConnectConfig, SFTPWrapper } from 'ssh2';
+import { Client } from 'ssh2';
+import { readFile } from 'fs/promises';
+import { debug } from './utils';
+
+export class SSHClient {
+  private sshConnection: Client;
+  private sftpConnection?: SFTPWrapper;
+
+  private connected = false;
+
+  constructor(private sshClientOptions: ConnectConfig) {
+    this.sshConnection = new Client();
+    this.setupEventListeners();
+  }
+
+  setupEventListeners() {
+    this.sshConnection.on('ready', () => {
+      debug('SSH: Connection established');
+      this.connected = true;
+    });
+    this.sshConnection.on('error', (err) => {
+      debug('SSH: Connection error', err);
+      this.connected = false;
+    });
+    this.sshConnection.on('close', () => {
+      debug('SSH: Connection closed');
+      this.connected = false;
+      this.sshConnection.destroy();
+    });
+  }
+
+  async connect() {
+    if (this.connected) {
+      return Promise.resolve();
+    }
+    const privateKey = this.sshClientOptions.privateKey
+      ? await readFile(this.sshClientOptions.privateKey)
+      : undefined;
+    return new Promise((resolve, reject) => {
+      this.sshConnection.connect({
+        ...this.sshClientOptions,
+        privateKey,
+      });
+      this.sshConnection.on('error', reject);
+      // @ts-expect-error We expect an error here - why?
+      this.sshConnection.on('ready', resolve);
```

Comment:
The "why" for this error is because TS cannot infer the type of the returned Promise here, so it falls back to `Promise<unknown>` and gives `resolve` a signature of `(value: unknown) => void` (i.e. one non-optional argument), which doesn't match the expected type for the callback passed to `.on('ready', cb)` (i.e. a function taking zero non-optional arguments), but since there are other events declared whose callbacks _do_ take arguments, TS gives a non-helpful error message thinking you passed the wrong event name.

There are a couple ways to go about this, the (imo) best one being to tell TS that you're returning a `Promise<void>` here; There is already a linter warning on this method and others about missing return types, just fixing those should also address this particular issue.

Reply:
That makes sense. Thanks for the explanation! I was really scratching my head as to why Typescript couldn't find the right overload 😁